### PR TITLE
Make arguments to engine optional

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
## Checklist

- [/] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [/] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [/] I have not broken the cheatsheet

### CODE REVIEW

**Comment by pokey** [comment]
update from meet-up re null tree-sitter:

- [ ] if tree-sitter is null, create DisabledLanguageDefinitions that just doesn't support any languages 
- [ ] remove tree-sitter debug prints
- [ ] allow null to pass to parse tree action for now and just have null check there that throws

goal for all this stuff is to allow undefined components to pass into engine, and contain all resulting null checks to `createEngine`, and create a `DisabledFoo` in response. Note that third checkbox above fails this criterion, but we don't need to be dogmatic

File: packages/cursorless-engine/src/core/HatTokenMapImpl.ts [hunk 1]
```diff
@@ -155,3 +155,24 @@ export class HatTokenMapImpl implements HatTokenMap {
     this.prePhraseMapsSnapshotTimestamp = hrtime.bigint();
   }
 }
+
+export class DisabledHatTokenMap implements HatTokenMap {
```
Comment (pokey) [inline packages/cursorless-engine/src/core/HatTokenMapImpl.ts:159]: move to own file

File: packages/cursorless-engine/src/cursorlessEngine.ts [hunk 2]
```diff
@@ -63,7 +73,10 @@ export async function createCursorlessEngine(
   const languageDefinitions = new LanguageDefinitions(fileSystem, treeSitter);
   await languageDefinitions.init();
 
-  const talonSpokenForms = new TalonSpokenFormsJsonReader(fileSystem);
+  const talonSpokenForms =
+    fileSystem != null
+      ? new TalonSpokenFormsJsonReader(fileSystem)
+      : new DisabledTalonSpokenFormsJsonReader();
```
Comment (pokey) [inline packages/cursorless-engine/src/cursorlessEngine.ts:79]: make sure not to merge this

File: packages/cursorless-engine/src/languages/LanguageDefinition.ts [hunk 3]
```diff
@@ -91,12 +88,12 @@ export class LanguageDefinition {
  * @returns The text of the query file, with all imports inlined
  */
 async function readQueryFileAndImports(
-  fileSystem: FileSystem,
-  languageQueryPath: string,
+  provider: LanguageDefinitionsProvider,
+  languageFilename: string,
```
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/languages/LanguageDefinition.ts:92]: The reason for switching from path to name is that if we have a non file system implementation of the language provider path is irrelevant. I think every definition name should just be unique and in which folder they are located doesn't really matter.

**Comment by AndreasArvidsson** [comment]
ready!

File: packages/cursorless-engine/src/languages/LanguageDefinition.ts [hunk 4]
```diff
@@ -136,10 +134,10 @@ async function readQueryFileAndImports(
       }
 
       if (doValidation) {
-        validateQueryCaptures(fileName, rawQuery);
+        validateQueryCaptures(filename, rawQuery);
       }
 
-      rawQueryStrings[queryPath] = rawQuery;
+      rawQueryStrings[filename] = rawQuery;
```
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/languages/LanguageDefinition.ts:140]: I'm tempted to simplify this regex. Remove support for `/` and don't support `include`

File: packages/cursorless-engine/src/languages/LanguageDefinition.ts [hunk 5]
```diff
@@ -91,43 +90,42 @@ export class LanguageDefinition {
...
-  fileSystem: FileSystem,
-  languageQueryPath: string,
+  ide: IDE,
+  provider: LanguageDefinitionsProvider,
+  languageQueryName: string,
 ) {
   // Seed the map with the query file itself
   const rawQueryStrings: Record<string, string | null> = {
-    [languageQueryPath]: null,
+    [languageQueryName]: null,
```
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/languages/LanguageDefinition.ts:99]: The reason for switching from path to name is that if we have a non file system implementation of the language provider, path is irrelevant. I think every definition name(scm filename) should just be unique and in which folder they are located doesn't really matter.

**Review by pokey — COMMENTED** [review]
wow this PR does a lot. I started in for a quick review but there's too much going on to easily see the picture. I would be tempted to split it into 4 PRs, but how about as a compromise let's just split it into 2: 

1. CommandServerApi / HatTokenMap
2. LanguageDefinitions / TreeSitter

I think that is a nice split point

See also the commit I just pushed, and comments below

File: packages/cursorless-engine/src/CommandHistory.ts [hunk 6]
```diff
@@ -27,7 +27,7 @@ export class CommandHistory implements CommandRunnerDecorator {
   constructor(
     private ide: IDE,
     private storage: CommandHistoryStorage,
-    private commandServerApi: CommandServerApi | null,
+    private commandServerApi: CommandServerApi | undefined,
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/CommandHistory.ts:30]: Unfortunately not. This is constructed in the extension file and does not get the disabled default implementation from the engine

File: packages/common/src/ide/types/LanguageDefinitionsProvider.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+export interface LanguageDefinitionsProvider {
+  /**
+   * Listen for changes to language definitions
+   */
+  onChanges(listener: () => void): Disposable;
+
+  /**
+   * Read a query definition. The query name is the name of one of our `.scm` files.
+   */
+  readQuery(name: string): Promise<string | undefined>;
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/ide/types/LanguageDefinitionsProvider.ts:12]: done

**Comment by AndreasArvidsson** [comment]
Ok this pull request has now been stripped off the Tree sitter stuff

**Review by pokey — COMMENTED** [review]
looks good with one minor comment!

File: packages/cursorless-engine/src/spokenForms/CustomSpokenForms.ts [hunk 8]
```diff
@@ -84,12 +84,11 @@ export class CustomSpokenForms {
         this.needsInitialTalonUpdate_ = true;
       } else {
         console.error("Error loading custom spoken forms", err);
+        const msg = (err as Error).message.replace(/\.$/, "");
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/spokenForms/CustomSpokenForms.ts:87]: Stripping duplication off trailing period

[the remark this answers is not in the document]
Comment (pokey) [reply packages/cursorless-engine/src/spokenForms/CustomSpokenForms.ts:87]: oh got it

**Review by pokey — APPROVED** [review]

**Pull request merged**